### PR TITLE
Dev app 1656 custom redirect block page

### PR DESCRIPTION
**Pull request by ilaipx**
[REFACTOR] added PxBlock class to handle blocking operations.
[ADD] custom block url with redirects.

### CODE REVIEW

File: PerimeterXModule/Internals/PxBlock.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+				pxContext.ApplicationContext.Response.Write(content);
+				return;
+			}
+
+			if (pxContext.BlockAction != "c" && pxContext.BlockAction != "b")
+			{
+				if (pxContext.BlockAction == "r")
+				{
+					pxContext.ApplicationContext.Response.StatusCode = 429; // HTTP/1.1 429 TooManyRequests
```
Comment (yaronpx) [inline PerimeterXModule/Internals/PxBlock.cs]: where do you load the rate limit template?

File: PerimeterXModule/Internals/PxBlock.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+					string uri = pxContext.ApplicationContext.Request.Url.AbsoluteUri;
+					string encodedUri = Convert.ToBase64String(Encoding.UTF8.GetBytes(uri));
+					string redirectUrl = string.Format("{0}?url={1}&uuid={2}&vid={3}", config.CustomBlockUrl, encodedUri, pxContext.UUID, pxContext.Vid);
+					PxLoggingUtils.LogDebug("Redirecting to custom block page: " + redirectUrl);
+					pxContext.ApplicationContext.Response.Redirect(redirectUrl);
+					return;
+				}
+
+				HttpResponseMessage response = httpClient.GetAsync(config.CustomBlockUrl).Result;
+				if ((int)response.StatusCode >= 300)
```
[the remark this answers is not in the document]
Comment (ilaipx) [reply PerimeterXModule/Internals/PxBlock.cs]: we fetch the html of the page, so 300 and up basically means we will not get it's html, am I 
correct?

File: PerimeterXModule/Internals/PxBlock.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+		public bool IsJsonResponse(PxContext pxContext)
+		{
+			Dictionary<string, string> headers = pxContext.GetHeadersAsDictionary();
+			string jsonHeader;
+			bool jsonHeaderExists = headers.TryGetValue("accept", out jsonHeader) || headers.TryGetValue("content-type", out jsonHeader);
+			if (jsonHeaderExists)
+			{
+				string[] values = jsonHeader.Split(',');
+				string value = Array.Find(values, v => v == "application/json");
```
[the remark this answers is not in the document]
Comment (ilaipx) [reply PerimeterXModule/Internals/PxBlock.cs]: thanks.

**Comment by yaronpx** [comment]
please fix conflicts

**Pull request merged**